sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java

```diff
@@ -204,25 +224,63 @@ public void after() {
     }
   }
 
-  @Test
-  @Category(NeedsRunner.class)
-  public void testTeardownCalledAfterExceptionInSetup() {
-    ExceptionThrowingOldFn fn = new ExceptionThrowingOldFn(MethodForException.SETUP);
-    p
-        .apply(Create.of(1, 2, 3))
-        .apply(ParDo.of(fn));
-    try {
-      p.run();
-      fail("Pipeline should have failed with an exception");
-    } catch (Exception e) {
-      assertThat(
-          "Function should have been torn down after exception",
-          ExceptionThrowingOldFn.teardownCalled.get(),
-          is(true));
+  private static class CallSequenceEnforcingStatefulFn<K, V> extends DoFn<KV<K, V>, KV<K, V>> {
```

Comment:
Despite the dangers of inheritance, I think it'd be cleaner to take CallSequenceEnforcingFn as a base than have another copy of the implementation.

Reply:
Done. Confirmed that the test still fails without the fix and succeeds with the fix, just for good measure.